File: functions/own_profile/get_my_friends.py
from firebase_admin import firestore
from models.constants import Collections, FriendshipFields, QueryOperators, Status
from models.data_models import FriendsResponse, Friend
from utils.logging_utils import get_logger


def get_my_friends(request) -> FriendsResponse:
    """
    Retrieves the current user's friends and pending friendship requests.

    This function fetches all accepted and pending friendships where the current user
    is in the members array, and returns the friend's information with status.

    Args:
        request: The Flask request object containing:
                - user_id: The authenticated user's ID (attached by authentication middleware)

    Returns:
        A FriendsResponse containing:
        - A list of Friend objects with the friend's profile information and friendship status
    """
    logger = get_logger(__name__)
    logger.info(f"Retrieving friends and pending requests for user: {request.user_id}")

    # Get the authenticated user ID from the request
    current_user_id = request.user_id

    # Initialize Firestore client
    db = firestore.client()

    # Use a single efficient query with array_contains and in operator for multiple statuses
    friendships_query = (
        db.collection(Collections.FRIENDSHIPS)
        .where(FriendshipFields.MEMBERS, QueryOperators.ARRAY_CONTAINS, current_user_id)
        .where(
            FriendshipFields.STATUS,
            QueryOperators.IN,
            [Status.ACCEPTED, Status.PENDING],
        )
    )

    logger.info(f"Querying friendships for user: {current_user_id}")

    # Execute the query
    friendships = list(friendships_query.stream())

    friends = []

    # Process friendships
    for doc in friendships:
        friendship_data = doc.to_dict()
        friendship_status = friendship_data.get(FriendshipFields.STATUS)

        # Determine if the current user is the sender or receiver
        if friendship_data.get(FriendshipFields.SENDER_ID) == current_user_id:
            # Current user is the sender, so friend is the receiver
            friend_id = friendship_data.get(FriendshipFields.RECEIVER_ID)
            friend_username = friendship_data.get(
                FriendshipFields.RECEIVER_USERNAME, ""
            )
            friend_name = friendship_data.get(FriendshipFields.RECEIVER_NAME, "")
            friend_avatar = friendship_data.get(FriendshipFields.RECEIVER_AVATAR, "")
        else:
            # Current user is the receiver, so friend is the sender
            friend_id = friendship_data.get(FriendshipFields.SENDER_ID)
            friend_username = friendship_data.get(FriendshipFields.SENDER_USERNAME, "")
            friend_name = friendship_data.get(FriendshipFields.SENDER_NAME, "")
            friend_avatar = friendship_data.get(FriendshipFields.SENDER_AVATAR, "")

        logger.info(
            f"Processing friendship with friend: {friend_id}, status: {friendship_status}"
        )

        friends.append(
            Friend(
                user_id=friend_id,
                username=friend_username,
                name=friend_name,
                avatar=friend_avatar,
            )
        )

    logger.info(
        f"Retrieved {len(friends)} friends and pending requests for user: {current_user_id}"
    )

    # Return the list of friends
    return FriendsResponse(friends=friends)

File: functions/src/own_profile/get-my-friends.ts
import { Request } from 'express';
import { getFirestore, QueryDocumentSnapshot } from 'firebase-admin/firestore';
import { ApiResponse, EventName, FriendEventParams } from '../models/analytics-events.js';
import { Collections, QueryOperators } from '../models/constants.js';
import { FriendDoc, friendConverter, ff } from '../models/firestore/friend-doc.js';
import { Friend, FriendsResponse, PaginationPayload } from '../models/data-models.js';
import { getLogger } from '../utils/logging-utils.js';
import { applyPagination, generateNextCursor, processQueryStream } from '../utils/pagination-utils.js';
import { formatTimestamp } from '../utils/timestamp-utils.js';

import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const logger = getLogger(path.basename(__filename));

/**
 * Retrieves the current user's friends and pending friendship requests.
 *
 * This function fetches all accepted and pending friendships where the current user
 * is in the member array and returns the friend's information with status.
 *
 * @param req - The Express request object containing:
 *              - userId: The authenticated user's ID (attached by authentication middleware)
 *              - validated_params: Pagination parameters containing:
 *                - limit: Maximum number of friends to return (default: 20, min: 1, max: 100)
 *                - after_cursor: Cursor for pagination (base64 encoded document path)
 *
 * @returns An ApiResponse containing the friend's response and analytics
 */
export const getMyFriends = async (req: Request): Promise<ApiResponse<FriendsResponse>> => {
  const db = getFirestore();
  const currentUserId = req.userId;

  logger.info(`Retrieving friends and pending requests for user: ${currentUserId}`);

  // Get pagination parameters from the validated request
  const validatedParams = req.validated_params as PaginationPayload;
  const limit = validatedParams?.limit || 20;
  const afterCursor = validatedParams?.after_cursor;

  logger.info(`Pagination parameters - limit: ${limit}, after_cursor: ${afterCursor}`);

  let query = db
    .collection(Collections.PROFILES)
    .doc(currentUserId)
    .collection(Collections.FRIENDS)
    .withConverter(friendConverter)
    .orderBy(ff('last_update_at'), QueryOperators.DESC);

  // Apply cursor-based pagination - Express will automatically catch errors
  const paginatedQuery = await applyPagination(query, afterCursor, limit);

  // Process friendships using streaming
  const { items: friendshipDocs, lastDoc } = await processQueryStream<QueryDocumentSnapshot>(
    paginatedQuery,
    (doc) => doc,
    limit,
  );

  const friends: Friend[] = [];

  // Process friendships
  for (const friendshipDoc of friendshipDocs) {
    const friendshipData = friendshipDoc.data() as FriendDoc;

    const friend: Friend = {
      user_id: friendshipDoc.id,
      username: friendshipData.username || '',
      name: friendshipData.name || '',
      avatar: friendshipData.avatar || '',
      last_update_emoji: friendshipData.last_update_emoji || '',
      last_update_time: (() => {
        const ts = friendshipData.last_update_at;
        return ts ? formatTimestamp(ts) : '';
      })(),
    };

    logger.info(`Processing friendship with friend: ${friend.user_id}`);

    friends.push(friend);
  }

  // Set up pagination for the next request
  const nextCursor = generateNextCursor(lastDoc, friends.length, limit);

  logger.info(`Retrieved ${friends.length} friends and pending requests for user: ${currentUserId}`);

  // Create analytics event
  const event: FriendEventParams = {
    friend_count: friends.length,
  };

  // Return the list of friends with pagination info
  const response: FriendsResponse = { friends, next_cursor: nextCursor };

  return {
    data: response,
    status: 200,
    analytics: {
      event: EventName.FRIENDS_VIEWED,
      userId: currentUserId,
      params: event,
    },
  };
};
